### PR TITLE
added `getsockname` for determining a socket name (address & port) of the running server

### PR DESCRIPTION
**Pull request by wildart**
I added `getsockname` for determining a socket name (address & port) of the running server. When server socket bound to `INPORT_ANY` port (0), during `listen` call the socket is automatically bound to a random free port. By calling  `getsockname`, you can get value of the assigned port.

In addition, I fixed a missing IPv6 support for `bind` and some documentation.

### CODE REVIEW

File: test/socket.jl [hunk 1]
```diff
@@ -174,3 +174,56 @@ begin
...
+    listen(sock)
+
+    new_addr, new_port = getsockname(sock)
+
+    @test default_addr == new_addr
+    @test default_port == new_port
+    close(sock)
+end
+
+@unix_only begin
```
Comment (tkelman) [inline test/socket.jl]: Any idea why this fails on windows? Implementation issue, or OS limitation?

File: test/socket.jl [hunk 2]
```diff
@@ -174,3 +174,56 @@ begin
...
+
+    @async begin
+        sleep(1)
+        ssock = connect(default_addr, default_port)
+    end
+
+    csock = accept(sock)
+    new_addr, new_port = getsockname(csock)
+
+    @test new_addr == IPv4("127.0.0.1")
```
Comment (vtjnash) [inline test/socket.jl]: travis notes that you bound this to "127.0.1.1" on the other end, so that's the reported address for the tests

File: src/jl_uv.c [hunk 3]
```diff
@@ -663,6 +663,46 @@ DLLEXPORT int jl_tcp_bind6(uv_tcp_t *handle, uint16_t port, void *host, unsigned
...
+    *family = addr.ss_family;
+    if (addr.ss_family == AF_INET) {
+        struct sockaddr_in* addr4 = (struct sockaddr_in*)&addr;
+        *port = addr4->sin_port;
+        memcpy(host, &(addr4->sin_addr), 4);
+    } else if (addr.ss_family == AF_INET6) {
+        struct sockaddr_in6* addr6 = (struct sockaddr_in6*)&addr;
+        *port = addr6->sin6_port;
+        memcpy(host, &(addr6->sin6_addr), 16);
+    }
```
Comment (vtjnash) [inline src/jl_uv.c]: else {
  return -1;
}

File: base/socket.jl [hunk 4]
```diff
@@ -723,3 +723,36 @@ function listenany(default_port)
...
+    rfamily = Ref(UInt32(0))
+    r = if isa(sock, TCPServer)
+        ccall(:jl_tcp_getsockname, Int32,
+                (Ptr{Void}, Ptr{UInt16}, Ptr{Void}, Ptr{UInt32}),
+                sock.handle, rport, raddress, rfamily)
+    else
+        ccall(:jl_tcp_getpeername, Int32,
+                (Ptr{Void}, Ptr{UInt16}, Ptr{Void}, Ptr{UInt32}),
+                sock.handle, rport, raddress, rfamily)
+    end
```
Comment (vtjnash) [inline base/socket.jl]: `uv_error("cannot obtain socket name", r);`

File: base/socket.jl [hunk 5]
```diff
@@ -723,3 +723,36 @@ function listenany(default_port)
...
+        port = ntoh(getindex(rport))
+        if getindex(rfamily) == 2 # AF_INET
+            addrv4 = raddress[1:4]
+            naddr = ntoh(unsafe_load(Ptr{UInt32}(pointer(addrv4)), 1))
+            addr = IPv4(naddr)
+        elseif getindex(rfamily) == 10 # AF_INET6
+            naddr = ntoh(unsafe_load(Ptr{UInt128}(pointer(raddress)), 1))
+            addr = IPv6(naddr)
+        else
+            warn("unsupported address family: $(getindex(rfamily))")
```
Comment (vtjnash) [inline base/socket.jl]: this should be an error

File: base/socket.jl [hunk 6]
```diff
@@ -723,3 +723,36 @@ function listenany(default_port)
...
+    r = if isa(sock, TCPServer)
+        ccall(:jl_tcp_getsockname, Int32,
+                (Ptr{Void}, Ptr{UInt16}, Ptr{Void}, Ptr{UInt32}),
+                sock.handle, rport, raddress, rfamily)
+    else
+        ccall(:jl_tcp_getpeername, Int32,
+                (Ptr{Void}, Ptr{UInt16}, Ptr{Void}, Ptr{UInt32}),
+                sock.handle, rport, raddress, rfamily)
+    end
+    addr = IPv4(UInt32(0))
```
Comment (vtjnash) [inline base/socket.jl]: no need to initialize these with values. `local addr` is sufficient

File: test/socket.jl (hunk 1 above)
Comment (wildart) [inline test/socket.jl]: Appveyor failed of this test on Windows, I guess they do not have IPv6 available on test machine. I'll turn if back. Problem is how to tell if IPv6 is on or not before executing this test. Or should I remove it?

Comment (ihnorton) [inline test/socket.jl]: see #9882 and #10335

File: test/socket.jl (hunk 2 above)
Comment (wildart) [inline test/socket.jl]: The point of this test to have different addresses for the server and the client. I bind server to  `127.0.1.1`, So when I call `getsockname` on a client socket, accepted by the server, the result should be an address of the connected client, assuming it is `localhost`. See netstat dump below: 
```
tcp        0      0 127.0.1.1:11011         0.0.0.0:*               LISTEN      24146/julia4    
tcp        0      0 127.0.1.1:11011         127.0.0.1:47940         ESTABLISHED 24146/julia4 
```
I pass this test on my machine. Unfortunately, travis thinks differently. Have any Idea?

**Comment by wildart** [comment]
Closed by mistake.

**Comment by amitmurthy** [comment]
@wildart , thanks for the PR. Please rebase against current master. While rebasing you will need to delete  https://github.com/JuliaLang/julia/blob/9c82059671379d7ee89fe2b7dc173678b97bd45f/src/jl_uv.c#L693-L704 and change the call in `managers.jl` accordingly.

**Comment by malmaud** [comment]
Since this won't make it into Julia .4, maybe we could carve out the essential part of this PR and move it to JuliaWeb now.

**Comment by wildart** [comment]
@malmaud You still need `uv_tcp_getsockname` call. Do you mean to address `libuv` directly?

**Comment by malmaud** [comment]
Ya, there is nothing technically blocking us from directly addressing libuv in a package, is there? It would just be temporary until the functionality is moved into base in .5.

**Comment by malmaud** [comment]
@wildart, want to rebase this? Would be cool to get it into .5.

**Comment by wildart** [comment]
Rebased and fixed errors, it's ready to merge.

**Comment by tkelman** [comment]
Doc  changes here need to be reflected in helpdb

**Comment by wildart** [comment]
Done.

File: doc/stdlib/base.rst [hunk 7]
```diff
@@ -893,11 +893,12 @@ System
 
    Get the local machine's host name.
 
-.. function:: getipaddr() -> AbstractString
+.. function:: getipaddr()
```
Comment (tkelman) [inline doc/stdlib/base.rst]: should this be `getipaddr() -> IPAddr`?

Comment (wildart) [inline doc/stdlib/base.rst]: Crap, it's so good to have multiple documentations.

**Comment by wildart** [comment]
Ready.

File: base/socket.jl [hunk 8]
```diff
@@ -732,3 +722,35 @@ function listenany(default_port)
...
+    r = if isa(sock, TCPServer)
+        ccall(:jl_tcp_getsockname, Int32,
+                (Ptr{Void}, Ref{Cushort}, Ptr{Void}, Ref{Cuint}),
+                sock.handle, rport, raddress, rfamily)
+    else
+        ccall(:jl_tcp_getpeername, Int32,
+                (Ptr{Void}, Ref{Cushort}, Ptr{Void}, Ref{Cuint}),
+                sock.handle, rport, raddress, rfamily)
+    end
+    uv_error("cannot obtain socket name", r);
```
Comment (ihnorton) [inline base/socket.jl]: Need to delete this.

Comment (ihnorton) [inline base/socket.jl]: Ah, nevermind.

File: base/docs/helpdb.jl [hunk 9]
```diff
@@ -12031,3 +12032,11 @@ Bitwise exclusive or
 ```
 """
 Base.(:$)(x, y)
+
+doc"""
+    getsockname(sock::Union{TCPServer, TCPSocket}) -> (IPAddr,UInt16)
+
+Get the IP address and the port of the TCP server socket, to which it is bound,
```
Comment (malmaud) [inline base/docs/helpdb.jl]: Extra comma

File: base/docs/helpdb.jl [hunk 10]
```diff
@@ -7463,9 +7463,10 @@ true
 all(p, itr)
 
 doc"""
-    bind(socket::Union{UDPSocket, TCPSocket}, host::IPv4, port::Integer)
+    bind(socket::Union{UDPSocket, TCPSocket}, host::IPAddr, port::Integer; ipv6only=false)
```
Comment (malmaud) [inline base/docs/helpdb.jl]: Is the IPv6 fix logically connected to creating getsockname? If not, maybe it should be a separate PR?

Comment (wildart) [inline base/docs/helpdb.jl]: Yes, I need IPv6 bind for testing IPv6 getsockname.

**Comment by malmaud** [comment]
Great, glad to finally have this functionality.
